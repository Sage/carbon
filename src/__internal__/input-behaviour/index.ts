export { InputBehaviour, InputContext } from "./input-behaviour.component";
export {
  InputGroupBehaviour,
  InputGroupContext,
} from "./input-group-behaviour.component";
export type { InputContextProps } from "./useInputBehaviour";
