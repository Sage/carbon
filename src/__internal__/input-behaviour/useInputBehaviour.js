import { useState, useRef, useCallback, useMemo } from "react";

const useInputBehaviour = (blockGroupBehaviour) => {
  const [hasFocus, setHasFocus] = useState(false);
  const [hasMouseOver, setHasMouseOver] = useState(false);

  const inputRef = useRef(null);

  const onFocus = useCallback(() => setHasFocus(true), []);

  const onBlur = useCallback(() => setHasFocus(false), []);

  const assignInput = useCallback((input) => {
    inputRef.current = input.current;
  }, []);

  // use mouse down rather than click to accommodate click and drag events too
  const onMouseDown = useCallback(() => {
    // use a zero timeout to ensure focus is applied even on click and drag events
    setTimeout(() => inputRef && inputRef.current && inputRef.current.focus());
  }, []);

  const onMouseEnter = useCallback(() => setHasMouseOver(true), []);

  const onMouseLeave = useCallback(() => setHasMouseOver(false), []);

  const contextValue = useMemo(
    () => ({
      hasFocus,
      hasMouseOver,
      onFocus: blockGroupBehaviour ? undefined : onFocus,
      onBlur: blockGroupBehaviour ? undefined : onBlur,
      onMouseDown,
      onMouseEnter: blockGroupBehaviour ? undefined : onMouseEnter,
      onMouseLeave: blockGroupBehaviour ? undefined : onMouseLeave,
      inputRef: assignInput,
    }),
    [
      hasFocus,
      hasMouseOver,
      onFocus,
      onBlur,
      onMouseDown,
      blockGroupBehaviour,
      onMouseEnter,
      onMouseLeave,
      assignInput,
    ]
  );

  return contextValue;
};

export default useInputBehaviour;
