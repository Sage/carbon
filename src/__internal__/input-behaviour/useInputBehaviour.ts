import { useState, useRef, useCallback, useMemo } from "react";

export interface InputContextProps {
  hasFocus?: boolean;
  hasMouseOver?: boolean;
  onFocus?: () => void;
  onBlur?: () => void;
  onMouseDown?: () => void;
  onMouseEnter?: () => void;
  onMouseLeave?: () => void;
  inputRef?: (input: {
    current: HTMLInputElement | HTMLTextAreaElement | null;
  }) => void;
}

const useInputBehaviour = (
  blockGroupBehaviour?: boolean
): InputContextProps => {
  const [hasFocus, setHasFocus] = useState(false);
  const [hasMouseOver, setHasMouseOver] = useState(false);

  const inputRef = useRef<HTMLInputElement | HTMLTextAreaElement | null>(null);

  const onFocus = useCallback(() => setHasFocus(true), []);

  const onBlur = useCallback(() => setHasFocus(false), []);

  const assignInput = useCallback(
    (input: { current: HTMLInputElement | HTMLTextAreaElement | null }) => {
      inputRef.current = input.current;
    },
    []
  );

  // use mouse down rather than click to accommodate click and drag events too
  const onMouseDown = useCallback(() => {
    requestAnimationFrame(() => {
      inputRef?.current?.focus({
        preventScroll: true,
      });
    });
  }, []);

  const onMouseEnter = useCallback(() => setHasMouseOver(true), []);

  const onMouseLeave = useCallback(() => setHasMouseOver(false), []);

  const contextValue = useMemo(
    () => ({
      hasFocus,
      hasMouseOver,
      onFocus: blockGroupBehaviour ? undefined : onFocus,
      onBlur: blockGroupBehaviour ? undefined : onBlur,
      onMouseDown,
      onMouseEnter: blockGroupBehaviour ? undefined : onMouseEnter,
      onMouseLeave: blockGroupBehaviour ? undefined : onMouseLeave,
      inputRef: assignInput,
    }),
    [
      hasFocus,
      hasMouseOver,
      onFocus,
      onBlur,
      onMouseDown,
      blockGroupBehaviour,
      onMouseEnter,
      onMouseLeave,
      assignInput,
    ]
  );

  return contextValue;
};

export default useInputBehaviour;
