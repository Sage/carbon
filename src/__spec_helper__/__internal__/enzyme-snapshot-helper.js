import toJson from "enzyme-to-json";

function snapshotWithoutProps(enzymeWrapper, propNames) {
  let propNameList = propNames;

  if (!Array.isArray(propNames)) {
    propNameList = [propNames];
  }

  const snapshotOptions = {
    map: (json) => {
      propNameList.forEach((prop) => {
        if (json.props[prop]) {
          json.props[prop] = `[ ${prop} object ]`;
        }
      });

      return json;
    },
  };

  return toJson(enzymeWrapper, snapshotOptions);
}

function noThemeSnapshot(enzymeWrapper) {
  return snapshotWithoutProps(enzymeWrapper, "theme");
}

export { snapshotWithoutProps, noThemeSnapshot };
