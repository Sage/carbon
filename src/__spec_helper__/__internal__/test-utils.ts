/* eslint-disable jest/no-conditional-expect */
/* eslint-disable jest/no-identical-title */
/* eslint-disable jest/no-export */
import { render } from "@testing-library/react";
import { sprintf } from "sprintf-js";
import {
  LayoutProps,
  FlexboxProps,
  BackgroundProps,
  PositionProps,
  GridProps,
} from "styled-system";

import { space } from "../../style/themes/base/base-theme.config";
import { mockMatchMedia } from "../mock-match-media";

const isUpper = (char: string) => char.toUpperCase() === char;
const humpToDash = (acc: string, char: string) =>
  `${acc}${isUpper(char) ? `-${char.toLowerCase()}` : char}`;

const toCSSCase = (str: string) => {
  return str.split("").reduce(humpToDash, "");
};

const assertStyleMatch = (
  styleSpec: { [key: string]: string | number | undefined },
  component: HTMLElement,
  opts?: jest.Options,
) => {
  Object.entries(styleSpec).forEach(([attr, value]) => {
    expect(component).toHaveStyleRule(toCSSCase(attr), value, opts);
  });
};

const makeArrayKeys = (n: number) => [...Array(n).keys()];

const marginProps = [
  ["m", "margin"],
  ["ml", "marginLeft"],
  ["mr", "marginRight"],
  ["mt", "marginTop"],
  ["mb", "marginBottom"],
  ["mx", "marginLeft"],
  ["mx", "marginRight"],
  ["my", "marginTop"],
  ["my", "marginBottom"],
] as const;

type MarginProps = {
  [K in (typeof marginProps)[number][0]]?: string | number;
};

const paddingProps = [
  ["p", "padding"],
  ["pl", "paddingLeft"],
  ["pr", "paddingRight"],
  ["pt", "paddingTop"],
  ["pb", "paddingBottom"],
  ["px", "paddingLeft"],
  ["px", "paddingRight"],
  ["py", "paddingTop"],
  ["py", "paddingBottom"],
] as const;

type PaddingProps = {
  [K in (typeof paddingProps)[number][0]]?: string | number;
};

const colorProps = [
  ["color", "color", "#CCCCCC"],
  ["bg", "background-color", "#FFFFFF"],
  ["opacity", "opacity", "0.5"],
] as const;

const widthProps = ["width", "width", "200px"] as const;

const heightProps = ["height", "height", "200px"] as const;

const layoutProps = [
  widthProps,
  heightProps,
  ["minWidth", "min-width", "120px"],
  ["maxWidth", "max-width", "120px"],
  ["minHeight", "min-height", "120px"],
  ["maxHeight", "max-height", "120px"],
  ["size", "width", "120px"],
  ["size", "height", "120px"],
  ["display", "display", "inline-block"],
  ["verticalAlign", "vertical-align", "baseline"],
  ["overflow", "overflow", "hidden"],
  ["overflowX", "overflow-x", "hidden"],
  ["overflowY", "overflow-y", "hidden"],
] as const;

const flexBoxProps = [
  ["alignItems", "alignItems", "center"],
  ["alignContent", "alignContent", "center"],
  ["justifyItems", "justifyItems", "center"],
  ["justifyContent", "justifyContent", "center"],
  ["flexWrap", "flexWrap", "wrap"],
  ["flexDirection", "flexDirection", "row-reverse"],
  ["flex", "flex", "1"],
  ["flexGrow", "flexGrow", "1"],
  ["flexShrink", "flexShrink", "1"],
  ["flexBasis", "flexBasis", "100px"],
  ["justifySelf", "justifySelf", "center"],
  ["alignSelf", "alignSelf", "center"],
  ["order", "order", "1"],
] as const;

const gridProps = [
  ["gridColumn", "gridColumn", "1 / span 2"],
  ["gridRow", "gridRow", "1 / span 2"],
  ["gridArea", "gridArea", "myArea"],
  ["gridAutoFlow", "gridAutoFlow", "column"],
  ["gridAutoRows", "gridAutoRows", "150px"],
  ["gridAutoColumns", "gridAutoColumns", "50px"],
  ["gridTemplateRows", "gridTemplateRows", "100px 300px"],
  ["gridTemplateColumns", "gridTemplateColumns", "auto auto auto auto"],
  ["gridTemplateAreas", "gridTemplateAreas", "myArea myArea . . ."],
] as const;

const backgroundProps = [
  [
    "background",
    "background",
    "lightblue url('foo.jpg') no-repeat fixed center",
  ],
  ["backgroundImage", "background-image", "url(foo.jpg)"],
  ["backgroundSize", "background-size", "center"],
  ["backgroundRepeat", "background-repeat", "no-repeat"],
] as const;

const positionProps = [
  ["top", "0px"],
  ["bottom", "0px"],
  ["right", "0px"],
  ["left", "0px"],
  ["position", "fixed"],
  ["position", "sticky"],
  ["position", "absolute"],
  ["position", "static"],
  ["position", "relative"],
] as const;

export const getDefaultValue = (value?: string | number) => {
  const spaceArrayLength = space.length - 1;
  if (value === undefined) return value;
  const parsedValue = +value;
  if (typeof value === "string" && parsedValue > spaceArrayLength) {
    return `${value}px`;
  }
  if (parsedValue <= spaceArrayLength) {
    return space[parsedValue];
  }
  if (parsedValue > spaceArrayLength) {
    return `${parsedValue * 8}px`;
  }
  return value;
};

const testStyledSystemGrid = (
  component: (gridProperties?: GridProps) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(gridProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, propName, value) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component(props));

        assertStyleMatch({ [propName]: value }, elementQuery());
      });
    },
  );
};

const testStyledSystemBackground = (
  component: (backgroundProperties?: BackgroundProps) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(backgroundProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, propName, value) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component({ ...props }));

        assertStyleMatch({ [styledSystemProp]: value }, elementQuery());
      });
    },
  );
};

const testStyledSystemPosition = (
  component: (positionProperties?: PositionProps) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(positionProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, value) => {
      it(`then ${styledSystemProp} should have been set correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component({ ...props }));

        assertStyleMatch({ [styledSystemProp]: value }, elementQuery());
      });
    },
  );
};

// this util will catch that a console output occurred without polluting the output when running the unit tests
const expectConsoleOutput = (
  message: string,
  type: "warn" | "error" = "error",
) => {
  if (!message) {
    throw new Error(`no ${type} message provided`);
  }

  expect.assertions(1);

  const consoleType = global.console[type];
  let consoleArgs: string[];

  jest.spyOn(global.console, type).mockImplementation((...args) => {
    if (!args.length) return;

    const msg = args.join(" ");
    const params = args.slice(1, args.length);

    if (sprintf(msg, ...params).includes(message)) {
      consoleArgs = args;
      return;
    }

    consoleType(...args);
  });

  return () => {
    if (consoleArgs) {
      // eslint-disable-next-line no-console
      expect(console[type]).toHaveBeenCalledWith(...consoleArgs);
    }

    global.console[type] = consoleType;
  };
};

const testStyledSystemMargin = (
  component: (spacingProps?: MarginProps) => JSX.Element,
  elementQuery: () => HTMLElement,
  defaults?: MarginProps,
  assertOpts?: jest.Options,
) => {
  describe("default props", () => {
    let StyleElement: HTMLElement;

    beforeAll(() => {
      render(component({ ...defaults }));
      StyleElement = elementQuery();
    });

    it("should set the correct margins", () => {
      let margin;
      let marginLeft;
      let marginRight;
      let marginTop;
      let marginBottom;

      if (defaults) {
        margin = getDefaultValue(defaults.m);
        marginLeft = getDefaultValue(defaults.ml || defaults.mx);
        marginRight = getDefaultValue(defaults.mr || defaults.mx);
        marginTop = getDefaultValue(defaults.mt || defaults.my);
        marginBottom = getDefaultValue(defaults.mb || defaults.my);

        assertStyleMatch(
          {
            margin,
            marginLeft,
            marginRight,
            marginTop,
            marginBottom,
          },
          StyleElement,
          assertOpts,
        );
      } else {
        expect(StyleElement).not.toHaveStyleRule("marginLeft");
        expect(StyleElement).not.toHaveStyleRule("marginRight");
        expect(StyleElement).not.toHaveStyleRule("marginTop");
        expect(StyleElement).not.toHaveStyleRule("marginBottom");
        expect(StyleElement).not.toHaveStyleRule("margin");
      }
    });
  });

  describe.each(marginProps)(
    'when a custom spacing is specified using the "%s" styled system props',
    (styledSystemProp, propName) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: 2 };
        render(component({ ...props }));

        assertStyleMatch(
          { [propName]: "var(--spacing200)" },
          elementQuery(),
          assertOpts,
        );
      });
    },
  );
};

const testStyledSystemPadding = (
  component: (spacingProps?: PaddingProps) => JSX.Element,
  elementQuery: () => HTMLElement,
  defaults?: PaddingProps,
  assertOpts?: jest.Options,
) => {
  describe("default props", () => {
    let StyleElement: HTMLElement;

    beforeAll(() => {
      render(component({ ...defaults }));
      StyleElement = elementQuery();
    });

    it("should set the correct paddings", () => {
      let padding;
      let paddingLeft;
      let paddingRight;
      let paddingTop;
      let paddingBottom;

      if (defaults) {
        padding = getDefaultValue(defaults.p);
        paddingLeft = getDefaultValue(defaults.pl || defaults.px);
        paddingRight = getDefaultValue(defaults.pr || defaults.px);
        paddingTop = getDefaultValue(defaults.pt || defaults.py);
        paddingBottom = getDefaultValue(defaults.pb || defaults.py);

        assertStyleMatch(
          {
            padding,
            paddingLeft,
            paddingRight,
            paddingTop,
            paddingBottom,
          },
          StyleElement,
          assertOpts,
        );
      } else {
        expect(StyleElement).not.toHaveStyleRule("paddingLeft");
        expect(StyleElement).not.toHaveStyleRule("paddingRight");
        expect(StyleElement).not.toHaveStyleRule("paddingTop");
        expect(StyleElement).not.toHaveStyleRule("paddingBottom");
        expect(StyleElement).not.toHaveStyleRule("padding");
      }
    });
  });

  describe.each(paddingProps)(
    'when a custom spacing is specified using the "%s" styled system props',
    (styledSystemProp, propName) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: 2 };
        render(component({ ...props }));

        assertStyleMatch(
          { [propName]: "var(--spacing200)" },
          elementQuery(),
          assertOpts,
        );
      });
    },
  );
};

const testStyledSystemSpacing = (
  component: (spacingProps?: MarginProps | PaddingProps) => JSX.Element,
  elementQuery: () => HTMLElement,
  defaults?: MarginProps | PaddingProps,
  assertOpts?: jest.Options,
) => {
  testStyledSystemMargin(
    component,
    elementQuery,
    defaults as MarginProps,
    assertOpts,
  );
  testStyledSystemPadding(
    component,
    elementQuery,
    defaults as PaddingProps,
    assertOpts,
  );
};

const testStyledSystemColor = (
  // https://stackoverflow.com/questions/53711454/styled-system-props-typing-with-typescript
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  component: (colorProperties?: any) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(colorProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, propName, value) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component({ ...props }));
        const StyleElement = elementQuery();
        // Some props need to have camelcase so used toHaveStyleRule rather than assertStyleMatch
        expect(StyleElement).toHaveStyleRule(propName, value);
      });
    },
  );
};

const testStyledSystemWidth = (
  component: (widthProperties?: { width: string }) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe("when a width prop is specified using styled system props", () => {
    it("should set the width styling correctly", () => {
      const [styledSystemProp, propName, value] = widthProps;
      const props = { [styledSystemProp]: value };
      render(component({ ...props }));
      const StyleElement = elementQuery();
      expect(StyleElement).toHaveStyleRule(propName, value);
    });
  });
};

const testStyledSystemHeight = (
  component: (heightProperties?: { height: string }) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe("when a height prop is specified using styled system props", () => {
    it("should set the height styling correctly", () => {
      const [styledSystemProp, propName, value] = heightProps;
      const props = { [styledSystemProp]: value };
      render(component({ ...props }));
      const StyleElement = elementQuery();
      expect(StyleElement).toHaveStyleRule(propName, value);
    });
  });
};

const testStyledSystemLayout = (
  component: (layoutProperties?: LayoutProps) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(layoutProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, propName, value) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component({ ...props }));
        const StyleElement = elementQuery();
        // Some props need to have camelcase so used toHaveStyleRule rather than assertStyleMatch
        expect(StyleElement).toHaveStyleRule(propName, value);
      });
    },
  );
};

const testStyledSystemFlexBox = (
  component: (flexboxProperties?: FlexboxProps) => JSX.Element,
  elementQuery: () => HTMLElement,
) => {
  describe.each(flexBoxProps)(
    'when a prop is specified using the "%s" styled system props',
    (styledSystemProp, propName, value) => {
      it(`should set ${propName} styling correctly`, () => {
        const props = { [styledSystemProp]: value };
        render(component(props));

        assertStyleMatch({ [propName]: value }, elementQuery());
      });
    },
  );
};

export {
  assertStyleMatch,
  toCSSCase,
  makeArrayKeys,
  mockMatchMedia,
  testStyledSystemColor,
  testStyledSystemLayout,
  testStyledSystemFlexBox,
  testStyledSystemGrid,
  testStyledSystemBackground,
  testStyledSystemPosition,
  expectConsoleOutput,
  testStyledSystemSpacing,
  testStyledSystemMargin,
  testStyledSystemPadding,
  testStyledSystemWidth,
  testStyledSystemHeight,
};
