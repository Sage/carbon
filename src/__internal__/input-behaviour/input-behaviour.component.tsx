import React from "react";
import useInputBehaviour, { InputContextProps } from "./useInputBehaviour";

const InputContext = React.createContext<InputContextProps>({});

export interface InputBehaviourProps {
  children?: React.ReactNode;
}

const InputBehaviour = ({ children }: InputBehaviourProps) => {
  const contextValue = useInputBehaviour();

  return (
    <InputContext.Provider value={{ ...contextValue }}>
      {children}
    </InputContext.Provider>
  );
};

export { InputContext, InputBehaviour };
