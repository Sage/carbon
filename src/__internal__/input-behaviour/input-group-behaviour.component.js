import React from "react";
import PropTypes from "prop-types";

import useInputBehaviour from "./useInputBehaviour";

const defaultValue = {};
const InputGroupContext = React.createContext(defaultValue);

const InputGroupBehaviour = ({ children, blockGroupBehaviour }) => {
  const contextValue = useInputBehaviour(blockGroupBehaviour);

  return (
    <InputGroupContext.Provider value={contextValue}>
      {children}
    </InputGroupContext.Provider>
  );
};

InputGroupBehaviour.propTypes = {
  children: PropTypes.node,
  blockGroupBehaviour: PropTypes.bool,
};

export { InputGroupContext, InputGroupBehaviour };
