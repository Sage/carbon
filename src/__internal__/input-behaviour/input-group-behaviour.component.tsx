import React from "react";

import useInputBehaviour, { InputContextProps } from "./useInputBehaviour";
import { InputBehaviourProps } from "./input-behaviour.component";

const InputGroupContext = React.createContext<InputContextProps>({});

export interface InputGroupBehaviourProps extends InputBehaviourProps {
  blockGroupBehaviour?: boolean;
}

const InputGroupBehaviour = ({
  children,
  blockGroupBehaviour,
}: InputGroupBehaviourProps) => {
  const contextValue = useInputBehaviour(blockGroupBehaviour);

  return (
    <InputGroupContext.Provider value={contextValue}>
      {children}
    </InputGroupContext.Provider>
  );
};

export { InputGroupContext, InputGroupBehaviour };
