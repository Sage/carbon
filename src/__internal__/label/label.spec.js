import React from "react";
import { shallow, mount } from "enzyme";

import Help from "../../components/help";
import Tooltip from "../../components/tooltip/tooltip.component";
import Label from ".";
import StyledLabel, { StyledLabelContainer } from "./label.style";
import { assertStyleMatch } from "../../__spec_helper__/test-utils";
import { noThemeSnapshot } from "../../__spec_helper__/enzyme-snapshot-helper";
import ValidationIcon from "../validations/validation-icon.component";
import mintTheme from "../../style/themes/mint";
import IconWrapperStyle from "./icon-wrapper.style";
import { InputContext, InputGroupContext } from "../input-behaviour";

jest.mock("@tippyjs/react/headless");

const validationTypes = ["error", "warning", "info"];

describe("Label", () => {
  it("renders the label", () => {
    expect(noThemeSnapshot(render({}, shallow))).toMatchSnapshot();
  });

  describe("context handlers", () => {
    let contextOnMouseEnter;
    let groupContextOnMouseEnter;

    let contextOnMouseLeave;
    let groupContextOnMouseLeave;

    let wrapper;

    beforeEach(() => {
      contextOnMouseEnter = jest.fn();
      groupContextOnMouseEnter = jest.fn();

      contextOnMouseLeave = jest.fn();
      groupContextOnMouseLeave = jest.fn();

      wrapper = renderWithContext(
        {},
        {
          onMouseEnter: groupContextOnMouseEnter,
          onMouseLeave: groupContextOnMouseLeave,
        },
        {
          onMouseEnter: contextOnMouseEnter,
          onMouseLeave: contextOnMouseLeave,
        }
      );
    });

    it("triggers onMouseEnter callbacks passed from context", () => {
      wrapper.find(StyledLabel).props().onMouseEnter();
      expect(contextOnMouseEnter).toHaveBeenCalled();
      expect(groupContextOnMouseEnter).toHaveBeenCalled();
    });

    it("triggers onMouseLeave callbacks passed from context", () => {
      wrapper.find(StyledLabel).props().onMouseLeave();
      expect(contextOnMouseLeave).toHaveBeenCalled();
      expect(groupContextOnMouseLeave).toHaveBeenCalled();
    });

    it("does nothing if onMouseEnter or onMouseLeave callbacks are not provided", () => {
      wrapper = render();
      const inputProps = wrapper.find(StyledLabel).props();
      inputProps.onMouseEnter();
      inputProps.onMouseLeave();
    });
  });

  describe("when initiated with the help prop", () => {
    it("contains Help component with the content specified in that prop", () => {
      const wrapper = render({ help: "Help me!" }, shallow);
      expect(wrapper.find(Help).contains("Help me!")).toBe(true);
    });

    it("passes tooltipId prop if provided", () => {
      const tooltipId = "tooltip-test";
      const wrapper = render({ help: "Help me!", tooltipId }, shallow);
      expect(wrapper.find(Help).props().tooltipId).toBe(tooltipId);
    });
  });

  describe("when inline", () => {
    it("applies styling for an inline label", () => {
      const wrapper = render({ inline: true });
      assertStyleMatch(
        {
          boxSizing: "border-box",
          marginBottom: "0",
          justifyContent: "flex-end",
          width: "30%",
        },
        wrapper.find(StyledLabelContainer)
      );
    });

    it("applies correct styling for inline label with 0 width", () => {
      const wrapper = render({ inline: true, width: 0 });
      assertStyleMatch(
        {
          width: "30%",
        },
        wrapper.find(StyledLabelContainer)
      );
    });

    it('applies styling for an inline "optional" label', () => {
      const wrapper = render({
        inline: true,
        optional: true,
        theme: mintTheme,
      });

      assertStyleMatch(
        {
          content: '"(optional)"',
          fontWeight: "350",
          marginLeft: "4px",
        },
        wrapper.find(StyledLabelContainer),
        { modifier: "::after" }
      );
    });

    it("applies styling when pr prop set", () => {
      assertStyleMatch(
        {
          paddingRight: "var(--spacing200)",
        },
        render({ inline: true, pr: 2 })
      );
    });

    it("applies styling when pl prop set", () => {
      assertStyleMatch(
        {
          paddingLeft: "var(--spacing200)",
        },
        render({ inline: true, pl: 2 })
      );
    });

    it('applies styling for an inline "isRequired" label', () => {
      const wrapper = render({
        inline: true,
        isRequired: true,
        theme: mintTheme,
      });

      assertStyleMatch(
        {
          content: '"*"',
          color: "var(--colorsSemanticNegative500)",
          fontWeight: "700",
          marginLeft: "var(--spacing100)",
        },
        wrapper.find(StyledLabel),
        { modifier: "::after" }
      );
    });
  });

  describe("when left aligned", () => {
    it("aligns the label correctly", () => {
      const wrapper = render({ inline: true, align: "left" });
      assertStyleMatch(
        {
          justifyContent: "flex-start",
        },
        wrapper.find(StyledLabelContainer)
      );
    });
  });

  describe("when disabled", () => {
    it("applies disabled color", () => {
      const wrapper = render({ disabled: true });

      assertStyleMatch(
        {
          color: "var(--colorsYin030)",
        },
        wrapper.find(StyledLabel)
      );
    });
  });

  describe("with disabled and error", () => {
    it("applies disabled color", () => {
      const wrapper = render({ error: true, disabled: true });

      assertStyleMatch(
        {
          color: "var(--colorsYin030)",
        },
        wrapper.find(StyledLabel)
      );
    });
  });

  describe("when the help icon is focused", () => {
    it("then the IconWrapper outline should have the expected value", () => {
      const wrapper = render({ help: "help message" }).find(IconWrapperStyle);
      wrapper.simulate("focus");

      assertStyleMatch(
        {
          outline: `2px solid var(--colorsSemanticFocus500)`,
        },
        wrapper,
        { modifier: ":focus" }
      );
    });
  });

  describe.each(validationTypes)(
    "when %s prop is passed as string",
    (vType) => {
      it("show validation icon", () => {
        const wrapper = render(
          { [vType]: "Message", useValidationIcon: true },
          mount
        );
        const icon = wrapper.find(ValidationIcon);

        expect(icon.exists()).toEqual(true);
      });

      it("passes tooltipId prop if provided", () => {
        const tooltipId = "tooltip-test";
        const wrapper = render(
          { [vType]: "Message", useValidationIcon: true, tooltipId },
          mount
        );
        expect(wrapper.find(ValidationIcon).props().tooltipId).toBe(tooltipId);
      });
    }
  );

  describe.each(validationTypes)(
    "when %s prop is passed as true boolean",
    (vType) => {
      it("do not show validation icon", () => {
        const wrapper = render(
          { [vType]: true, useValidationIcon: true },
          mount
        );
        const icon = wrapper.find(ValidationIcon);

        expect(icon.exists()).toEqual(false);
      });
    }
  );

  describe.each(validationTypes)(
    "when %s prop is passed as string and label is inline",
    (vType) => {
      it('tooltip should render in position "top"', () => {
        const wrapper = render(
          { [vType]: "Message", useValidationIcon: true, inline: true },
          mount
        );
        const icon = wrapper.find(ValidationIcon);
        expect(icon.prop("tooltipPosition")).toBe("top");
      });
    }
  );

  describe.each(validationTypes)(
    "when %s prop is passed as string and label is not inline",
    (vType) => {
      it('tooltip should render in position "right"', () => {
        const wrapper = render(
          { [vType]: "Message", useValidationIcon: true },
          mount
        );
        const icon = wrapper.find(ValidationIcon);
        expect(icon.prop("tooltipPosition")).toBe("right");
      });
    }
  );

  describe("when help prop is passed", () => {
    const help = "help me!";

    it("consumed focus flag controls visibility prop of Tooltip", () => {
      const hasFocus = true;
      const wrapper = renderWithContext(
        { help },
        {},
        {
          hasFocus,
        }
      );

      const tooltipProps = wrapper.find(Tooltip).props();
      const desiredTooltipProps = {
        message: help,
        isVisible: hasFocus,
      };
      expect(tooltipProps).toMatchObject(desiredTooltipProps);
    });

    it("consumed mouse over flag controls visibility prop of Tooltip", () => {
      const hasMouseOver = true;
      const wrapper = renderWithContext(
        { help },
        {},
        {
          hasMouseOver,
        }
      );

      const tooltipProps = wrapper.find(Tooltip).props();
      const desiredTooltipProps = {
        message: help,
        isVisible: hasMouseOver,
      };
      expect(tooltipProps).toMatchObject(desiredTooltipProps);
    });
  });
});

function render(props, renderer = mount) {
  return renderer(<Label {...props}>Name:</Label>);
}

function renderWithContext(
  props = {},
  inputGroupContextValue = {},
  inputContextValue = {}
) {
  return mount(
    <InputGroupContext.Provider value={inputGroupContextValue}>
      <InputContext.Provider value={inputContextValue}>
        <Label {...props}>Name:</Label>
      </InputContext.Provider>
    </InputGroupContext.Provider>
  );
}
