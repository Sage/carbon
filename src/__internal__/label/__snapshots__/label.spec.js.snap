// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label renders the label 1`] = `
<styled.div
  align="right"
  width={30}
>
  <styled.label
    data-element="label"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
  >
    Name:
  </styled.label>
</styled.div>
`;
